source/common/network/listen_socket_impl.cc

```diff
@@ -36,28 +36,28 @@ void ListenSocketImpl::setListenSocketOptions(const Network::Socket::OptionsShar
   }
 }
 
-TcpListenSocket::TcpListenSocket(const Address::InstanceConstSharedPtr& address,
-                                 const Network::Socket::OptionsSharedPtr& options,
-                                 bool bind_to_port)
-    : ListenSocketImpl(address->socket(Address::SocketType::Stream), address) {
-  RELEASE_ASSERT(fd_ != -1, "");
-
-  // TODO(htuch): This might benefit from moving to SocketOptionImpl.
-  int on = 1;
-  int rc = setsockopt(fd_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
-  RELEASE_ASSERT(rc != -1, "");
-
+void ListenSocketImpl::setupSocket(const Network::Socket::OptionsSharedPtr& options,
+                                   bool bind_to_port) {
   setListenSocketOptions(options);
 
   if (bind_to_port) {
     doBind();
   }
 }
 
-TcpListenSocket::TcpListenSocket(int fd, const Address::InstanceConstSharedPtr& address,
-                                 const Network::Socket::OptionsSharedPtr& options)
-    : ListenSocketImpl(fd, address) {
-  setListenSocketOptions(options);
+template <>
+void NetworkListenSocket<
+    NetworkSocketTrait<Address::SocketType::Stream>>::setProtocolSpecificSocketOptions() {
+  // TODO(htuch): This might benefit from moving to SocketOptionImpl.
+  int on = 1;
+  int rc = setsockopt(fd_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
+  RELEASE_ASSERT(rc != -1, "");
+}
+
+template <>
+void NetworkListenSocket<
+    NetworkSocketTrait<Address::SocketType::Datagram>>::setProtocolSpecificSocketOptions() {
+  // TODO (Jojy): Do we need any UDP specific socket options?
```

Comment:
My vote would be not to set any socket options for UDP here, per [this comment](https://github.com/mpwarres/envoy/blob/57cdb58baec49d0315574854d0e8add9251e544f/source/common/network/listen_socket_impl.cc#L69) from #5115. Users that want to tune UDP socket options can always do so via SocketOptionImpl.

Reply:
I think to be consistent with the TCP code path, we should leave this here. Later when we move TCP's `SO_REUSEADDR` setup, we can completely remove the `setProtocolSpecificSocketOptions` method. What do you think?

Reply:
Sorry, I wasn't very clear. I agree it makes sense to keep the method, I just think that currently it should be left empty.  i.e. I think the current form is fine, was just weighing in on the TODO comment.

Reply:
Yeah, agree, if we can't come up with anything we can just remove the TODO here.